### PR TITLE
Add 3d tools libraries

### PR DESCRIPTION
**Pull request by scottander**
Adds 3d tools library w/

- Measurement3d widget
- Daylight widget
- Elevation Profile widget
- Line of sight widget
- Shadow cast widget
- Slice widget

### CODE REVIEW

File: libraries/3d-tools/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Esri 3d Widget Components
+
+Esri's JavaScript 4.x API contains a series of ready-to-use UI widgets. Included in this library are some of Esri's 3d widgets wrapped in a VertiGIS Studio Web Web component so that they can be used in a VertiGIS Studio Web application.
```
Comment (connerleverett) [inline libraries/3d-tools/README.md:3]: Just confirming that `Studio Web Web Component` is intentional

File: libraries/3d-tools/package.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    "name": "library-3d-tools",
+    "version": "0.0.0",
+    "private": true,
+    "scripts": {
+        "build": "vertigis-web-sdk build",
+        "start": "vertigis-web-sdk start"
+    },
+    "dependencies": {
+        "@arcgis/core": "^4.22.2",
+        "@vertigis/web": "5.16.0-rc.14825",
```
Comment (connerleverett) [inline libraries/3d-tools/package.json:11]: Should this just be `5.1.6.0`?

File: libraries/3d-tools/src/locale/inv.json [hunk 3]
```diff
@@ -0,0 +1,45 @@
+{
+    "language-web-incubator-area-measurement-3d-title": "Area Measurement",
+    "language-web-incubator-line-of-sight-3d-title": "Line of Sight",
+    "language-web-incubator-daylight-widget-3d-title": "Daylight",
+    "language-web-incubator-elevation-profile-3d-title": "Elevation Profile",
+    "language-web-incubator-shadow-cast-3d-title": "Shadow Cast",
+    "language-web-incubator-slice-3d-title": "Slice",
+    "language-web-incubator-area-measurement-3d-description": "Area Measurement",
+    "language-web-incubator-line-of-sight-3d-description": "Line of Sight",
```
Comment (connerleverett) [inline libraries/3d-tools/src/locale/inv.json:9]: This could be a better description.

**Review by connerleverett — CHANGES_REQUESTED** [review]

File: libraries/3d-tools/README.md (hunk 1 above)
Comment (scottander) [reply libraries/3d-tools/README.md]: Nope.

[the remark this answers is not in the document]
Comment (scottander) [reply libraries/3d-tools/README.md]: Updated

File: libraries/3d-tools/package.json (hunk 2 above)
Comment (scottander) [reply libraries/3d-tools/package.json:11]: I will update this when web is released

File: libraries/3d-tools/src/components/AreaMeasurement/AreaMeasurement.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import { ReactElement, useState, useEffect } from "react";
+import {
+    createEsriMapWidget,
+    MapWidgetProps,
+} from "@vertigis/web/ui/esriUtils";
+import AreaMeasurement3DWidget from "@arcgis/core/widgets/AreaMeasurement3D";
+import type Accessor from "@arcgis/core/core/Accessor";
+import AreaMeasurementModel from "./AreaMeasurementModel";
+
+export type DaylightWidgetProps = MapWidgetProps<
```
[the remark this answers is not in the document]
Comment (scottander) [reply libraries/3d-tools/src/components/AreaMeasurement/AreaMeasurement.tsx]: Updated

File: libraries/3d-tools/src/components/ElevationProfile/ElevationProfile.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+>(ElevationProfileWidget, false, true);
+
+export default function ElevationProfile(
+    props: ElevationProfileWidgetProps
+): ReactElement {
+    const { model } = props;
+    const { map, regionService } = model;
+    const [widget, setWidget] = useState<ElevationProfileWidget>();
+    useWatchAndRerender(map, "map");
+    useWatchAndRerender(model, [
```
[the remark this answers is not in the document]
Comment (scottander) [reply libraries/3d-tools/src/components/ElevationProfile/ElevationProfile.tsx]: Updated

File: libraries/3d-tools/src/locale/inv.json (hunk 3 above)
Comment (scottander) [reply libraries/3d-tools/src/locale/inv.json]: Updated

File: libraries/3d-tools/src/locale/inv.json [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    "language-web-incubator-daylight-widget-3d-description": "The daylight widget can be used to manipulate the time and date and to toggle shadows in a SceneView.",
+    "language-web-incubator-elevation-profile-3d-description": "Elevation Profile",
+    "language-web-incubator-shadow-cast-3d-description": "Shadow Cast",
```
[the remark this answers is not in the document]
Comment (scottander) [reply libraries/3d-tools/src/locale/inv.json]: Updated

**Review by connerleverett — APPROVED** [review]

File: libraries/3d-tools/app/layout.xml [hunk 7]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<layout xmlns="https://geocortex.com/layout/v1" xmlns:custom="vertigis.web.incubator">
```
Comment (jazhall) [inline libraries/3d-tools/app/layout.xml:2]: Let's change the alias used here to be xmlns:incubator instead. This is one of the suggested changes made in the Time slider widget PR.

File: libraries/3d-tools/package.json (hunk 2 above)
Comment (eric-parsons) [inline libraries/3d-tools/package.json:11]: This will need to change to the final version.

File: libraries/3d-tools/package.json [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+        "mapillary-js": "^4.0.0"
```
Comment (eric-parsons) [inline libraries/3d-tools/package.json:12]: Don't think that this is supposed to be here.

File: libraries/3d-tools/package.json [hunk 9]
```diff
@@ -0,0 +1,18 @@
+{
+    "name": "library-3d-tools",
+    "version": "0.0.0",
```
Comment (eric-parsons) [inline libraries/3d-tools/package.json:3]: It might make sense to give this a proper version?

**Review by eric-parsons — APPROVED** [review]

File: libraries/3d-tools/app/layout.xml (hunk 7 above)
Comment (scottander) [reply libraries/3d-tools/app/layout.xml]: updated

File: libraries/3d-tools/src/components/Daylight/DaylightModel.ts [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    ComponentModelProperties,
+    importModel,
+} from "@vertigis/web/models";
+
+interface DaylightModelProperties extends ComponentModelProperties {
+    playButtons?: boolean;
+    shadowsToggle?: boolean;
+    datePicker?: boolean;
+    timezone?: boolean;
+    dateOrSeason?: "date" | "season";
```
[the remark this answers is not in the document]
Comment (scottander) [reply libraries/3d-tools/src/components/Daylight/DaylightModel.ts]: updated

File: libraries/3d-tools/package.json (hunk 2 above)
Comment (scottander) [reply libraries/3d-tools/package.json]: updated to final versions

File: libraries/3d-tools/package.json (hunk 8 above)
Comment (scottander) [reply libraries/3d-tools/package.json]: nope, removed :)

**Review by jazhall — APPROVED** [review]

**Comment by pairinmason** [comment]
**Question1**
On 3D Tools startup page, can we make the toolbar scrolled all the way to the left and display the currently active tab (i.e. Area Measurement)
image

**Comment by pairinmason** [comment]
**Question2**
The following links are not working testing in the branch. Are they meant to be working after merged to incubator master?
image

**Comment by pairinmason** [comment]
**Question3**
`Measurement3d widget`, do we need to include `measure distance` as well?

**Comment by pairinmason** [comment]
**Issue4**
Clicking `New measurement` button give console error.
image

**Comment by pairinmason** [comment]
**Issue5**
`Elevation Profile` tab needs scrollbar to access the rest of the form when:
>select `Elevation Profile` tab
expand `Ground`

**Comment by pairinmason** [comment]
**Question6**
`Elevation Profile`, can you help guiding on how to use the `Select line` button?
I click the button and follow the instruction by clicking in the scene. It doesn't perform anything.
image

**Comment by pairinmason** [comment]
**Question7**
`Elevation Profile` , beside `Ground ` should we also include `View`?
image

**Comment by pairinmason** [comment]
**Issue8**
`Shadow Cast`, clicking outside of color picker flyout doesn't close it.
image

**Comment by pairinmason** [comment]
**Issue9**
`Shadow Cast`, the text size of `Visualization dropdown list` is too small.
image

**Comment by pairinmason** [comment]
**Issue10**
`Shadow Cast`, when selecting mode of `Visualization `, the width of form and text size shrink.
image

**Comment by pairinmason** [comment]
**Question12**
`Slice`, unable to select to `Exclude layer` from the slice.
 image

**Comment by pairinmason** [comment]
**Question13**
Re-confirming that we don't include `3D Sketch` as part of `incubator 3D Tools`? 
image

**Comment by pairinmason** [comment]
**Issue15**
In `Web`, `Elevation Profile`, 
>https://apps.vertigisstudio.com/web/?app=4f90981ef33942a1aaee32b330818263
Perform `Area Measurement` using 3D tool in tab

Error in console:
image

**Comment by pairinmason** [comment]
**Issue15**
In `Web`, `Slice`,
>https://apps.vertigisstudio.com/web/?app=4f90981ef33942a1aaee32b330818263
Click `New slice` button
Click on map

Error in console:
image

**Comment by pairinmason** [comment]
**Issue16**
In Web, Slice,

https://apps.vertigisstudio.com/web/?app=4f90981ef33942a1aaee32b330818263
Click New slice button
Add slice on map
Click `Exclude layer` button
Click on a layer

Error:
image

**Comment by pairinmason** [comment]
**Issue17**
`Elevation Profile`, width shrink in `Firefox`
image

**Comment by pairinmason** [comment]
**Issue18**
`Shadow Cast`, the text size of these 2 buttons are still small.
image

**Comment by pairinmason** [comment]
**Issue19**
>1.https://webzoo.latitudegeo.com/builds/AppDesigner/master/?webUrl=https://webzoo.latitudegeo.com/builds/Web/2070/
2.Upload 3D Tools library (from [http://localhost:3000/main.js](http://localhost:3000/main.js) )in Web Designer  
3.Drag 3D tools components into Tabs in layout

No 3D tools appearing in tab in Web preview.

image

**Comment by pairinmason** [comment]
Issue20
In `Web`, When I select `Slice tab` in web with/without adding a slice and without clicking Exclude Layer, clicking on a feature on map giving error and failing to display results-details.
Need to reload the app to get `map.click` to work again

Error in console:
image

**Comment by pairinmason** [comment]
**Note21**
 If `Line of Sight `or `Slice `is active on app startup. The app won't load in 5.16.1 designer.
 This is because of the Esri weird layers, we fix it in the related arc-x PR (merged into 5.17 Web).
 
 
image

**Review by pairinmason — APPROVED** [review]

**Pull request merged**